File: scripts/verify-plugin-parity.mjs
#!/usr/bin/env node
// Proves the Claude plugin's discoverable surface actually resolves, and stamps
// a structural digest so a packaged copy cannot silently differ from its version.
//
// The failure this prevents was concrete: an installed plugin cache frozen at an
// older layout (pre-`src/`) shared version 0.1.0-dev.0 with a working tree whose
// manifest pointed the MCP server at a path the cache did not contain. Nothing
// signalled the drift because the version string was identical.
//
// The surface a Claude plugin exposes, by convention plus manifest:
//   skills/<id>/SKILL.md          capabilities and explicit entrypoints
//   agents/<name>.md              authority agents
//   .claude-plugin/plugin.json    mcpServers → the legion MCP server entry point
//   hooks/hooks.json              lifecycle hook command targets
//
// Every declared artifact must resolve on disk. The digest is computed over the
// surface's shape and the referenced entry points' relative paths — not their
// full contents — so a bump is required whenever the discoverable structure
// changes, which is exactly the invariant version numbers must carry.
import { readFileSync, writeFileSync, readdirSync, existsSync, statSync } from 'node:fs';
import { delimiter, resolve, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { createHash } from 'node:crypto';

const ROOT = resolve(fileURLToPath(new URL('..', import.meta.url)));
const SURFACE_FILE = 'src/registry/plugin-surface.json';
const DISTRIBUTION_CONTRACT = 'release/distribution-contract.json';
const CHANNELS_FILE = 'packaging/channels.json';

const readJson = (p) => JSON.parse(readFileSync(p, 'utf8'));
const ACTIVATION_PREFLIGHT = 'node scripts/verify-plugin-parity.mjs --check';

// Resolve a ${CLAUDE_PLUGIN_ROOT}-relative reference to a repo path.
const pluginRel = (ref) => ref.replace(`\${CLAUDE_PLUGIN_ROOT}/`, '');

function bareCommand(command) {
  if (typeof command !== 'string') return null;
  const value = command.trim();
  // A command with an argument, path, or shell interpolation is not a PATH
  // binary declaration. Path-relative targets are checked by the existing
  // hook-target check below.
  return /^[A-Za-z0-9][A-Za-z0-9._-]*$/.test(value) ? value : null;
}

function executableOnPath(command, env = process.env) {
  const pathValue = env.PATH ?? env.Path ?? '';
  const entries = pathValue.split(delimiter);
  const extensions = process.platform === 'win32'
    ? ['', ...(env.PATHEXT ?? '.COM;.EXE;.BAT;.CMD').split(';').filter(Boolean)]
    : [''];
  const candidates = [...new Set(extensions.map((extension) => `${command}${extension}`))];

  for (const entry of entries) {
    for (const candidate of candidates) {
      const path = resolve(entry || process.cwd(), candidate);
      try {
        const stat = statSync(path);
        // Windows resolves executable extensions without a Unix executable bit.
        if (stat.isFile() && (process.platform === 'win32' || (stat.mode & 0o111) !== 0)) return path;
      } catch { /* this PATH entry does not provide the binary */ }
    }
  }
  return null;
}

export function activeBootstrap(root) {
  let contract;
  let channels;
  try { contract = readJson(join(root, DISTRIBUTION_CONTRACT)); } catch { /* checked below */ }
  try { channels = readJson(join(root, CHANNELS_FILE)); } catch { /* checked below */ }

  const nativeRelease = contract?.nativeRelease;
  const directChannel = channels?.channels?.['direct-bootstrap'];
  if (nativeRelease?.status !== 'available' || directChannel?.status !== 'available') return null;

  const contractUrl = contract?.nativeRelease?.bootstrapAuthority;
  const channelUrl = channels?.bootstrap?.stableUrl;
  const directUrl = directChannel?.stableUrl;
  if (
    typeof contractUrl !== 'string'
    || contractUrl !== channelUrl
    || contractUrl !== directUrl
  ) return null;
  return contractUrl;
}

function declaredPathBinaries(manifest, hooks) {
  const declarations = new Map();
  const add = (command, source) => {
    const binary = bareCommand(command);
    if (!binary) return;
    if (!declarations.has(binary)) declarations.set(binary, []);
    declarations.get(binary).push(source);
  };

  for (const [id, server] of Object.entries(manifest.mcpServers ?? {})) {
    add(server?.command, `.claude-plugin/plugin.json mcpServers.${id}`);
  }
  for (const [event, entries] of Object.entries(hooks.hooks ?? {})) {
    for (const entry of entries ?? []) {
      for (const [index, hook] of (entry.hooks ?? []).entries()) {
        add(hook?.command, `hooks/hooks.json ${event}[${index}]`);
      }
    }
  }
  return declarations;
}

/**
 * Check the external binaries that the plugin invokes by bare name. This is
 * deliberately a preflight: a package-manager manifest can document an
 * optional alias, but it cannot make a missing PATH command runnable. The
 * canonical recovery is therefore always named in the failure itself.
 */
export function checkPathBinaries(root = ROOT, { manifest, hooks, env = process.env } = {}) {
  const pluginManifest = manifest ?? readJson(join(root, '.claude-plugin', 'plugin.json'));
  const hookConfig = hooks ?? readJson(join(root, 'hooks', 'hooks.json'));
  const bootstrap = activeBootstrap(root);
  const binaries = [];
  const problems = [];

  for (const [binary, sources] of declaredPathBinaries(pluginManifest, hookConfig)) {
    const resolved = executableOnPath(binary, env);
    const record = { binary, sources, resolved };
    binaries.push(record);
    if (!resolved && !bootstrap) {
      const bootstrapNote = `Install via the published direct bootstrap, then rerun '${ACTIVATION_PREFLIGHT}'.`;
      problems.push(
        `plugin binary '${binary}' is not reachable on PATH (required by ${sources.join(', ')}). ${bootstrapNote}`,
      );
    }
  }

  return { binaries, problems };
}

export function collectSurface(root = ROOT, { checkInstalledBinaries = true } = {}) {
  const problems = [];

  // Skills — every subdirectory that ships a SKILL.md. _shared and manifests are
  // support directories and legitimately carry none.
  const skillsDir = join(root, 'skills');
  const skills = readdirSync(skillsDir)
    .filter((id) => { try { return statSync(join(skillsDir, id)).isDirectory(); } catch { return false; } })
    .filter((id) => existsSync(join(skillsDir, id, 'SKILL.md')))
    .sort();

  // Agents — every markdown file under agents/, keyed by frontmatter name.
  const agentsDir = join(root, 'agents');
  const agents = existsSync(agentsDir)
    ? readdirSync(agentsDir).filter((f) => f.endsWith('.md')).sort().map((f) => {
        const text = readFileSync(join(agentsDir, f), 'utf8');
        const name = /^name:\s*(.+)$/m.exec(text)?.[1]?.trim() ?? null;
        if (!name) problems.push(`agent ${f} has no frontmatter name`);
        return { file: `agents/${f}`, name };
      })
    : [];

  // MCP servers — source-backed entries must resolve. Installed-native entries
  // must use Legion's exact stdio launch contract.
  const manifest = readJson(join(root, '.claude-plugin', 'plugin.json'));
  const mcpServers = Object.entries(manifest.mcpServers ?? {}).map(([id, server]) => {
    const args = server.args ?? [];
    const native = server.command === 'legion' && JSON.stringify(args) === JSON.stringify(['serve', '--stdio']);
    const entry = args.map(pluginRel).find((a) => a.endsWith('.mjs') || a.endsWith('.js'));
    if (!native && !entry) problems.push(`mcp server ${id} declares neither installed native Legion nor a resolvable entry point`);
    else if (entry && !existsSync(join(root, entry))) problems.push(`mcp server ${id} entry point missing: ${entry}`);
    return { id, command: server.command ?? null, args, entry: entry ?? null };
  });

  // Hooks — every command target referenced by hooks.json must resolve.
  const hooks = readJson(join(root, 'hooks', 'hooks.json'));
  const hookEvents = Object.keys(hooks.hooks ?? {}).sort();
  const hookTargets = new Set();
  for (const entries of Object.values(hooks.hooks ?? {})) {
    for (const entry of entries) {
      for (const hook of entry.hooks ?? []) {
        const m = /\$\{CLAUDE_PLUGIN_ROOT\}\/([^\s"]+)/.exec(hook.command ?? '');
        if (m) hookTargets.add(m[1]);
      }
    }
  }
  for (const target of hookTargets) {
    if (!existsSync(join(root, target))) problems.push(`hook command target missing: ${target}`);
  }

  if (checkInstalledBinaries) problems.push(...checkPathBinaries(root, { manifest, hooks }).problems);

  return {
    version: manifest.version,
    skills,
    agents,
    mcpServers,
    hooks: { events: hookEvents, targets: [...hookTargets].sort() },
    problems,
  };
}

// Digest over structure and entry-point paths, NOT file contents: a bump is
// required when the discoverable shape changes, not on every content edit.
export function surfaceDigest(surface) {
  const shape = {
    skills: surface.skills,
    agents: surface.agents.map((a) => a.name),
    mcpServers: surface.mcpServers.map((s) => ({ id: s.id, command: s.command, args: s.args, entry: s.entry })),
    hooks: surface.hooks,
  };
  return `sha256:${createHash('sha256').update(JSON.stringify(shape)).digest('hex')}`;
}

export function buildSurfaceRecord(root = ROOT, options = {}) {
  const surface = collectSurface(root, options);
  return {
    schemaVersion: 1,
    kind: 'legion-plugin-surface',
    version: surface.version,
    digest: surfaceDigest(surface),
    counts: {
      skills: surface.skills.length,
      agents: surface.agents.length,
      mcpServers: surface.mcpServers.length,
      hookEvents: surface.hooks.events.length,
    },
    surface: { skills: surface.skills, agents: surface.agents, mcpServers: surface.mcpServers, hooks: surface.hooks },
    problems: surface.problems,
  };
}

const isMain = process.argv[1] && resolve(process.argv[1]) === resolve(fileURLToPath(import.meta.url));
if (isMain) {
  // Repository integrity is environment-independent. Installed PATH health is
  // checked by plugin:verify plus clean-environment release qualification.
  const record = buildSurfaceRecord(ROOT, {
    checkInstalledBinaries: !process.argv.includes('--structural-only'),
  });
  const target = join(ROOT, SURFACE_FILE);
  const rendered = `${JSON.stringify(record, null, 2)}\n`;

  if (record.problems.length) {
    process.stderr.write(`plugin surface does not resolve:\n  - ${record.problems.join('\n  - ')}\n`);
    process.exit(1);
  }

  if (process.argv.includes('--check')) {
    const current = existsSync(target) ? readJson(target) : null;
    if (!current) { process.stderr.write(`missing ${SURFACE_FILE}; run: node scripts/verify-plugin-parity.mjs\n`); process.exit(1); }
    // The lifecycle invariant: a changed structural digest REQUIRES a changed
    // version. Same digest + same version is fine; changed digest + same version
    // is the exact drift the packaged cache suffered.
    if (current.digest !== record.digest && current.version === record.version) {
      process.stderr.write(
        `plugin surface changed but version did not.\n` +
        `  recorded ${current.digest} @ ${current.version}\n` +
        `  current  ${record.digest} @ ${record.version}\n` +
        `Bump the version in .claude-plugin/plugin.json (and package.json), then rerun.\n`);
      process.exit(1);
    }
    if (JSON.stringify(current) !== JSON.stringify(record)) {
      process.stderr.write(`${SURFACE_FILE} is stale; run: node scripts/verify-plugin-parity.mjs\n`);
      process.exit(1);
    }
    process.stdout.write(`plugin surface: resolves, ${record.counts.skills} skills / ${record.counts.agents} agents / ${record.counts.mcpServers} mcp / ${record.counts.hookEvents} hook events, digest ${record.digest.slice(0, 19)}…\n`);
  } else {
    writeFileSync(target, rendered);
    process.stdout.write(`wrote ${SURFACE_FILE}: ${record.counts.skills} skills, ${record.counts.agents} agents, ${record.counts.mcpServers} mcp, ${record.counts.hookEvents} hook events\n`);
  }
}

File: engine/bins/legion-dev/src/generators/verify_plugin_parity.rs
//! Port of `scripts/verify-plugin-parity.mjs`.

use serde_json::{json, Map, Value};
use sha2::{Digest, Sha256};
use std::collections::BTreeSet;
use std::fs;
use std::path::Path;

const SURFACE_FILE: &str = "src/registry/plugin-surface.json";
const DISTRIBUTION_CONTRACT: &str = "release/distribution-contract.json";
const CHANNELS_FILE: &str = "packaging/channels.json";
const ACTIVATION_PREFLIGHT: &str = "node scripts/verify-plugin-parity.mjs --check";

fn read_json(path: &Path) -> Result<Value, String> {
    let text = fs::read_to_string(path).map_err(|e| format!("{}: {e}", path.display()))?;
    serde_json::from_str(&text).map_err(|e| format!("{}: {e}", path.display()))
}

fn plugin_rel(reference: &str) -> String {
    reference.replace("${CLAUDE_PLUGIN_ROOT}/", "")
}

fn bare_command(command: Option<&str>) -> Option<String> {
    let value = command?.trim();
    let ok = !value.is_empty()
        && value.as_bytes()[0].is_ascii_alphanumeric()
        && value.bytes().all(|b| (b as char).is_ascii_alphanumeric() || b == b'.' || b == b'_' || b == b'-');
    if ok { Some(value.to_string()) } else { None }
}

fn executable_on_path(command: &str) -> Option<String> {
    let path_value = std::env::var("PATH").unwrap_or_default();
    let entries: Vec<&str> = path_value.split(if cfg!(windows) { ';' } else { ':' }).collect();
    let extensions: Vec<String> = if cfg!(windows) {
        let pathext = std::env::var("PATHEXT").unwrap_or_else(|_| ".COM;.EXE;.BAT;.CMD".to_string());
        std::iter::once(String::new()).chain(pathext.split(';').filter(|s| !s.is_empty()).map(String::from)).collect()
    } else {
        vec![String::new()]
    };
    let mut seen = BTreeSet::new();
    let candidates: Vec<String> = extensions.into_iter().map(|ext| format!("{command}{ext}")).filter(|c| seen.insert(c.clone())).collect();

    let cwd = std::env::current_dir().unwrap_or_default();
    for entry in entries {
        for candidate in &candidates {
            let base = if entry.is_empty() { cwd.clone() } else { std::path::PathBuf::from(entry) };
            let path = base.join(candidate);
            if let Ok(meta) = fs::metadata(&path) {
                if meta.is_file() {
                    #[cfg(unix)]
                    {
                        use std::os::unix::fs::PermissionsExt;
                        if meta.permissions().mode() & 0o111 != 0 {
                            return Some(path.display().to_string());
                        }
                    }
                    #[cfg(windows)]
                    {
                        return Some(path.display().to_string());
                    }
                }
            }
        }
    }
    None
}

/// Faithful port of `activeBootstrap(root)`.
pub fn active_bootstrap(root: &Path) -> Option<String> {
    let contract = read_json(&root.join(DISTRIBUTION_CONTRACT)).ok();
    let channels = read_json(&root.join(CHANNELS_FILE)).ok();

    let native_release = contract.as_ref().and_then(|c| c.get("nativeRelease"));
    let direct_channel = channels.as_ref().and_then(|c| c.get("channels")).and_then(|c| c.get("direct-bootstrap"));

    if native_release.and_then(|n| n.get("status")).and_then(Value::as_str) != Some("available")
        || direct_channel.and_then(|d| d.get("status")).and_then(Value::as_str) != Some("available")
    {
        return None;
    }

    let contract_url = native_release.and_then(|n| n.get("bootstrapAuthority")).and_then(Value::as_str);
    let channel_url = channels.as_ref().and_then(|c| c.get("bootstrap")).and_then(|b| b.get("stableUrl")).and_then(Value::as_str);
    let direct_url = direct_channel.and_then(|d| d.get("stableUrl")).and_then(Value::as_str);

    match contract_url {
        Some(c) if Some(c) == channel_url && Some(c) == direct_url => Some(c.to_string()),
        _ => None,
    }
}

fn declared_path_binaries(manifest: &Value, hooks: &Value) -> Vec<(String, Vec<String>)> {
    let mut order: Vec<String> = Vec::new();
    let mut map: std::collections::HashMap<String, Vec<String>> = std::collections::HashMap::new();
    let mut add = |command: Option<&str>, source: String| {
        if let Some(binary) = bare_command(command) {
            if !map.contains_key(&binary) {
                order.push(binary.clone());
            }
            map.entry(binary).or_default().push(source);
        }
    };
    if let Some(servers) = manifest.get("mcpServers").and_then(Value::as_object) {
        for (id, server) in servers {
            add(server.get("command").and_then(Value::as_str), format!(".claude-plugin/plugin.json mcpServers.{id}"));
        }
    }
    if let Some(events) = hooks.get("hooks").and_then(Value::as_object) {
        for (event, entries) in events {
            for entry in entries.as_array().into_iter().flatten() {
                for (index, hook) in entry.get("hooks").and_then(Value::as_array).into_iter().flatten().enumerate() {
                    add(hook.get("command").and_then(Value::as_str), format!("hooks/hooks.json {event}[{index}]"));
                }
            }
        }
    }
    order.into_iter().map(|b| { let s = map.remove(&b).unwrap_or_default(); (b, s) }).collect()
}

pub struct PathBinaryReport {
    pub problems: Vec<String>,
}

pub fn check_path_binaries(root: &Path, manifest: &Value, hooks: &Value) -> PathBinaryReport {
    let bootstrap = active_bootstrap(root);
    let mut problems = Vec::new();
    for (binary, sources) in declared_path_binaries(manifest, hooks) {
        let resolved = executable_on_path(&binary);
        if resolved.is_none() && bootstrap.is_none() {
            let bootstrap_note = format!("Install via the published direct bootstrap, then rerun '{ACTIVATION_PREFLIGHT}'.");
            problems.push(format!(
                "plugin binary '{binary}' is not reachable on PATH (required by {}). {bootstrap_note}",
                sources.join(", ")
            ));
        }
    }
    PathBinaryReport { problems }
}

pub struct Surface {
    pub version: Value,
    pub skills: Vec<String>,
    pub agents: Vec<Value>,
    pub mcp_servers: Vec<Value>,
    pub hook_events: Vec<String>,
    pub hook_targets: Vec<String>,
    pub problems: Vec<String>,
}

pub fn collect_surface(root: &Path, check_installed_binaries: bool) -> Result<Surface, String> {
    let mut problems = Vec::new();

    let skills_dir = root.join("skills");
    let mut skills: Vec<String> = fs::read_dir(&skills_dir)
        .map_err(|e| e.to_string())?
        .filter_map(|e| e.ok())
        .filter(|e| e.path().is_dir())
        .filter_map(|e| e.file_name().into_string().ok())
        .filter(|id| skills_dir.join(id).join("SKILL.md").is_file())
        .collect();
    skills.sort();

    let agents_dir = root.join("agents");
    let mut agents = Vec::new();
    if agents_dir.is_dir() {
        let mut files: Vec<String> = fs::read_dir(&agents_dir)
            .map_err(|e| e.to_string())?
            .filter_map(|e| e.ok())
            .filter_map(|e| e.file_name().into_string().ok())
            .filter(|f| f.ends_with(".md"))
            .collect();
        files.sort();
        for f in files {
            let text = fs::read_to_string(agents_dir.join(&f)).map_err(|e| e.to_string())?;
            let name = text
                .lines()
                .find_map(|l| l.strip_prefix("name:").map(|v| v.trim().to_string()))
                .filter(|s| !s.is_empty());
            if name.is_none() {
                problems.push(format!("agent {f} has no frontmatter name"));
            }
            let mut o = Map::new();
            o.insert("file".into(), Value::from(format!("agents/{f}")));
            o.insert("name".into(), name.map(Value::from).unwrap_or(Value::Null));
            agents.push(Value::Object(o));
        }
    }

    let manifest = read_json(&root.join(".claude-plugin/plugin.json"))?;
    let mut mcp_servers = Vec::new();
    if let Some(servers) = manifest.get("mcpServers").and_then(Value::as_object) {
        for (id, server) in servers {
            let args: Vec<Value> = server.get("args").and_then(Value::as_array).cloned().unwrap_or_default();
            let native = server.get("command").and_then(Value::as_str) == Some("legion")
                && args == vec![Value::from("serve"), Value::from("--stdio")];
            let entry = args
                .iter()
                .filter_map(Value::as_str)
                .map(plugin_rel)
                .find(|a| a.ends_with(".mjs") || a.ends_with(".js"));
            if !native && entry.is_none() {
                problems.push(format!("mcp server {id} declares neither installed native Legion nor a resolvable entry point"));
            } else if let Some(e) = &entry {
                if !root.join(e).exists() {
                    problems.push(format!("mcp server {id} entry point missing: {e}"));
                }
            }
            let mut o = Map::new();
            o.insert("id".into(), Value::from(id.clone()));
            o.insert("command".into(), server.get("command").cloned().unwrap_or(Value::Null));
            o.insert("args".into(), Value::Array(args));
            o.insert("entry".into(), entry.map(Value::from).unwrap_or(Value::Null));
            mcp_servers.push(Value::Object(o));
        }
    }

    let hooks = read_json(&root.join("hooks/hooks.json"))?;
    let mut hook_events: Vec<String> = hooks.get("hooks").and_then(Value::as_object).map(|o| o.keys().cloned().collect()).unwrap_or_default();
    hook_events.sort();
    let mut hook_targets: BTreeSet<String> = BTreeSet::new();
    if let Some(events) = hooks.get("hooks").and_then(Value::as_object) {
        for entries in events.values() {
            for entry in entries.as_array().into_iter().flatten() {
                for hook in entry.get("hooks").and_then(Value::as_array).into_iter().flatten() {
                    if let Some(cmd) = hook.get("command").and_then(Value::as_str) {
                        if let Some(idx) = cmd.find("${CLAUDE_PLUGIN_ROOT}/") {
                            let rest = &cmd[idx + "${CLAUDE_PLUGIN_ROOT}/".len()..];
                            let target: String = rest.chars().take_while(|c| !c.is_whitespace() && *c != '"').collect();
                            if !target.is_empty() {
                                hook_targets.insert(target);
                            }
                        }
                    }
                }
            }
        }
    }
    for target in &hook_targets {
        if !root.join(target).exists() {
            problems.push(format!("hook command target missing: {target}"));
        }
    }

    if check_installed_binaries {
        problems.extend(check_path_binaries(root, &manifest, &hooks).problems);
    }

    Ok(Surface {
        version: manifest.get("version").cloned().unwrap_or(Value::Null),
        skills,
        agents,
        mcp_servers,
        hook_events,
        hook_targets: hook_targets.into_iter().collect(),
        problems,
    })
}

/// Faithful port of `surfaceDigest(surface)`. `JSON.stringify` (default,
/// no whitespace) is matched by `serde_json::to_string` (compact).
pub fn surface_digest(surface: &Surface) -> String {
    let agent_names: Vec<Value> = surface.agents.iter().map(|a| a.get("name").cloned().unwrap_or(Value::Null)).collect();
    let mcp_shape: Vec<Value> = surface
        .mcp_servers
        .iter()
        .map(|s| {
            json!({
                "id": s.get("id"),
                "command": s.get("command"),
                "args": s.get("args"),
                "entry": s.get("entry"),
            })
        })
        .collect();
    let shape = json!({
        "skills": surface.skills,
        "agents": agent_names,
        "mcpServers": mcp_shape,
        "hooks": { "events": surface.hook_events, "targets": surface.hook_targets },
    });
    let mut hasher = Sha256::new();
    hasher.update(serde_json::to_string(&shape).unwrap().as_bytes());
    format!("sha256:{}", hex::encode(hasher.finalize()))
}

pub fn build_surface_record(root: &Path, check_installed_binaries: bool) -> Result<Value, String> {
    let surface = collect_surface(root, check_installed_binaries)?;
    let digest = surface_digest(&surface);
    let mut counts = Map::new();
    counts.insert("skills".into(), Value::from(surface.skills.len()));
    counts.insert("agents".into(), Value::from(surface.agents.len()));
    counts.insert("mcpServers".into(), Value::from(surface.mcp_servers.len()));
    counts.insert("hookEvents".into(), Value::from(surface.hook_events.len()));

    let mut hooks_out = Map::new();
    hooks_out.insert("events".into(), Value::Array(surface.hook_events.iter().cloned().map(Value::from).collect()));
    hooks_out.insert("targets".into(), Value::Array(surface.hook_targets.iter().cloned().map(Value::from).collect()));

    let mut surface_out = Map::new();
    surface_out.insert("skills".into(), Value::Array(surface.skills.iter().cloned().map(Value::from).collect()));
    surface_out.insert("agents".into(), Value::Array(surface.agents.clone()));
    surface_out.insert("mcpServers".into(), Value::Array(surface.mcp_servers.clone()));
    surface_out.insert("hooks".into(), Value::Object(hooks_out));

    let mut record = Map::new();
    record.insert("schemaVersion".into(), Value::from(1));
    record.insert("kind".into(), Value::from("legion-plugin-surface"));
    record.insert("version".into(), surface.version.clone());
    record.insert("digest".into(), Value::from(digest));
    record.insert("counts".into(), Value::Object(counts));
    record.insert("surface".into(), Value::Object(surface_out));
    record.insert("problems".into(), Value::Array(surface.problems.iter().cloned().map(Value::from).collect()));
    Ok(Value::Object(record))
}

pub fn run(root: &Path, check: bool) -> bool {
    run_opts(root, check, false)
}

pub fn run_opts(root: &Path, check: bool, structural_only: bool) -> bool {
    let record = match build_surface_record(root, !structural_only) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("verify-plugin-parity: {e}");
            return false;
        }
    };
    let problems = record.get("problems").and_then(Value::as_array).cloned().unwrap_or_default();
    let target = root.join(SURFACE_FILE);
    let rendered = format!("{}\n", serde_json::to_string_pretty(&record).unwrap());

    if !problems.is_empty() {
        let lines: Vec<String> = problems.iter().filter_map(|p| p.as_str().map(String::from)).collect();
        eprintln!("plugin surface does not resolve:\n  - {}", lines.join("\n  - "));
        return false;
    }

    if check {
        let current: Option<Value> = fs::read_to_string(&target).ok().and_then(|s| serde_json::from_str(&s).ok());
        let current = match current {
            Some(c) => c,
            None => {
                eprintln!("missing {SURFACE_FILE}; run: node scripts/verify-plugin-parity.mjs");
                return false;
            }
        };
        let current_digest = current.get("digest").and_then(Value::as_str).unwrap_or_default();
        let record_digest = record.get("digest").and_then(Value::as_str).unwrap_or_default();
        let current_version = current.get("version").cloned().unwrap_or(Value::Null);
        let record_version = record.get("version").cloned().unwrap_or(Value::Null);
        if current_digest != record_digest && current_version == record_version {
            eprintln!(
                "plugin surface changed but version did not.\n  recorded {current_digest} @ {current_version}\n  current  {record_digest} @ {record_version}\nBump the version in .claude-plugin/plugin.json (and package.json), then rerun."
            );
            return false;
        }
        if current != record {
            eprintln!("{SURFACE_FILE} is stale; run: node scripts/verify-plugin-parity.mjs");
            return false;
        }
        let counts = &record["counts"];
        println!(
            "plugin surface: resolves, {} skills / {} agents / {} mcp / {} hook events, digest {}…",
            counts["skills"], counts["agents"], counts["mcpServers"], counts["hookEvents"],
            &record_digest.chars().take(19).collect::<String>()
        );
        return true;
    }

    if let Err(e) = fs::write(&target, &rendered) {
        eprintln!("verify-plugin-parity: {e}");
        return false;
    }
    let counts = &record["counts"];
    println!(
        "wrote {SURFACE_FILE}: {} skills, {} agents, {} mcp, {} hook events",
        counts["skills"], counts["agents"], counts["mcpServers"], counts["hookEvents"]
    );
    true
}
